fix #920: Create recipes

zakpatterson opened the pull request:
<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

**What kind of change does this PR introduce?** (check at least one)
<!--
If you are introducing a new binding, you must reference an issue where this binding has been proposed, discussed and approved by the maintainers.
-->

- [ ] Bugfix
- [X] Feature
- [ ] New Binding Issue #___
- [ ] Code style update
- [x] Refactor
- [X] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)
<!--
If yes, please describe the impact and migration path for existing applications in an attached issue. Filing a PR with breaking changes that has not been discussed and approved by the maintainers in an issue will be immediately closed.
-->

- [ ] Yes. Issue #___
- [X] No


**The PR fulfills these requirements:**

- [X] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [X] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
fixes #920 

A recipe consists of:
	* Name
	* List of NPM dependencies required (dev and production)
	* Sub-commands needed
		For example, the react recipes run create-react-app to up the UI. Future recipes can do anything, they could pull down a sample UI, clone a repository, download a static HTML file, initialize another environment that compiles to JS/wasm etc.

The devPath and dist folders nomally configured by the init command are overridden by the recipe definition.

Example process for initializing a tauri project with the React + Typescript recipe:

![2020-08-07-103915_606x175_scrot](https://user-images.githubusercontent.com/1987109/89663863-dbf4b780-d89b-11ea-8d02-c2ac1d742e40.png)

Complete output (with some yarn output removed)
```
/home/projects/recipe-example
⟩ yarn tauri init
yarn run v1.22.4
$ /home/projects/recipe-example/node_modules/.bin/tauri init
[tauri]: running init
? What is your app name? Recipe Example
? What should the window title be? Recipe Example
? Would you like to add a UI recipe? React with Typescript
 dependency:manager Installing missing dependencies... +0ms
 dependency:cargo-commands "tauri-bundler" is already installed +14ms
 app:spawn [sync] Running "cargo generate-lockfile" +2ms

    Updating crates.io index
 dependency:crates "tauri" is already installed +889ms
 dependency:npm-packages "tauri" is already installed +4s
 recipe:install Installing dependencies for React with Typescript +0ms
 dependency:npm-packages Installing typescript... +4s
 app:spawn [sync] Running "yarn add typescript" +0ms

[ yarn install elided]

 dependency:npm-packages Installing @types/node... +10s
 app:spawn [sync] Running "yarn add @types/node" +0ms

[yarn install elided] 

 dependency:npm-packages Installing @types/react... +9s
 app:spawn [sync] Running "yarn add @types/react" +0ms

[yarn install elided] 

 dependency:npm-packages Installing @types/react-dom... +10s
 app:spawn [sync] Running "yarn add @types/react-dom" +0ms

[yarn install elided] 

 dependency:npm-packages Installing @types/jest... +10s
 app:spawn [sync] Running "yarn add @types/jest" +0ms

[yarn install elided] 

 dependency:npm-packages Installing create-react-app... +11s
 app:spawn [sync] Running "yarn add create-react-app --dev" +0ms

[yarn install elided] 

 recipe:postconfig Running post configuration for React with Typescript +6s
 app:spawn [sync] Running "yarn create-react-app --template typescript app-ui" +0ms

$ /home/projects/recipe-example/node_modules/.bin/create-react-app --template typescript app-ui

Creating a new React app in /home/projects/recipe-example/app-ui.

Installing packages. This might take a couple of minutes.
Installing react, react-dom, and react-scripts with cra-template-typescript...

[create-react-app yarn deps elided]

Initialized a git repository.

Installing template dependencies using yarnpkg...

[ Create react app yarn install elided]

We detected TypeScript in your project (src/App.test.tsx) and created a tsconfig.json file for you.

Your tsconfig.json has been populated with default values.

Removing template package using yarnpkg...

[1/2] Removing module cra-template-typescript...
[2/2] Regenerating lockfile and installing missing dependencies...
info fsevents@2.1.2: The platform "linux" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.12: The platform "linux" is incompatible with this module.
info "fsevents@1.2.12" is an optional dependency and failed compatibility check. Excluding it from installation.
warning " > @testing-library/user-event@7.2.1" has unmet peer dependency "@testing-library/dom@>=5".
success Uninstalled packages.

Created git commit.

Success! Created app-ui at /home/projects/recipe-example/app-ui
Inside that directory, you can run several commands:

  yarn start
    Starts the development server.

  yarn build
    Bundles the app into static files for production.

  yarn test
    Starts the test runner.

  yarn eject
    Removes this tool and copies build dependencies, configuration files
    and scripts into the app directory. If you do this, you can’t go back!

We suggest that you begin by typing:

  cd app-ui
  yarn start

Happy hacking!
 react-recipe 
  Your installation completed.
  Run these commands to start:
    For development server (provides UI): cd app-ui; yarn start
    For tauri (runs desktop app) in this directory: yarn tauri dev  
 +17s
Done in 113.14s.
```

zakpatterson committed:
Create recipes

A recipe consists of:
	* Name
	* List of NPM dependencies required (dev and production)
	* Sub-commands needed
		For example, the react recipes run create-react-app to set
		up the UI. Future recipes can do anything, they could pull
		down a sample UI, clone a repository, download a static HTML
		file, initialize another environment that compiles to JS/wasm etc.

The devPath and dist folders nomally configured by the init command
are overridden by the recipe definition.

zakpatterson committed:
Fix no recipe prompts

zakpatterson committed:
Fix linter unsafe assignment error

jbolda commented:
Notes as I have been reviewing and trying it out:
- This is part of the init command. Should we add another top level command that builds on top of the init command?
- I ran it as a `yarn global` (via `yarn global add $PWD` in the `./cli/tauri.js` folder), and it seems to use yarn but also returned a npm lockfile. This may be related to CRA?
- I think it makes sense to add scripts that make it easier to run after the process is done. A readme may also help with instructions on what to do next.
- Do we need to have opinions on yarn versus npm? We may want to encourage one versus the other, but I did also notice a mix of lock files so each init script may have an built in assumption as well.
- We are going to need opinions on folder structure. I think tauri assumes a `src-tauri` with the "app" code, and `src` with the UI / web code. 

```shell
❯ tauri init
[tauri]: running init
? What is your app name? test1
? What should the window title be? Tauri App
? Would you like to add a UI recipe? React.js
 dependency:manager Installing missing dependencies... +0ms
 dependency:cargo-commands "tauri-bundler" is already installed +123ms
 app:spawn [sync] Running "cargo generate-lockfile" +2ms

    Updating crates.io index
 dependency:crates "tauri" is already installed +359ms
 dependency:npm-packages Installing tauri... +3s
 app:spawn [sync] Running "npm install tauri" +1ms

npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142

> sharp@0.25.4 install C:\Users\Jacob\Documents\dev\test1\node_modules\sharp
> (node install/libvips && node install/dll-copy && prebuild-install --runtime=napi) || (node-gyp rebuild && node install/dll-copy)

info sharp Using cached C:\Users\Jacob\AppData\Roaming\npm-cache\_libvips\libvips-8.9.1-win32-x64.tar.gz
info sharp Creating C:\Users\Jacob\Documents\dev\test1\node_modules\sharp\build\Release
info sharp Copying DLLs from C:\Users\Jacob\Documents\dev\test1\node_modules\sharp\vendor\lib to C:\Users\Jacob\Documents\dev\test1\node_modules\sharp\build\Release

> optipng-bin@7.0.0 postinstall C:\Users\Jacob\Documents\dev\test1\node_modules\optipng-bin
> node lib/install.js

  √ optipng pre-build test passed successfully

> pngquant-bin@6.0.0 postinstall C:\Users\Jacob\Documents\dev\test1\node_modules\pngquant-bin
> node lib/install.js

  √ pngquant pre-build test passed successfully

> zopflipng-bin@6.0.1 postinstall C:\Users\Jacob\Documents\dev\test1\node_modules\zopflipng-bin
> node lib/install.js

  √ zopflipng pre-build test passed successfully
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.1.2 (node_modules\chokidar\node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
npm WARN test1@1.0.0 No description
npm WARN test1@1.0.0 No repository field.

+ tauri@0.11.1
added 673 packages from 505 contributors and audited 674 packages in 45.818s

45 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

 recipe:install Installing dependencies for React.js +0ms
 dependency:npm-packages Installing react... +3s
 app:spawn [sync] Running "npm install react" +1ms

npm WARN test1@1.0.0 No description
npm WARN test1@1.0.0 No repository field.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.1.3 (node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

+ react@16.13.1
added 5 packages from 2 contributors and audited 679 packages in 3.078s

45 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

 dependency:npm-packages Installing create-react-app... +6s
 app:spawn [sync] Running "npm install create-react-app --save-dev" +1ms

npm WARN test1@1.0.0 No description
npm WARN test1@1.0.0 No repository field.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.1.3 (node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

+ create-react-app@3.4.1
added 42 packages from 20 contributors and audited 721 packages in 5.249s

45 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

 recipe:postconfig Running post configuration for React.js +6s
 app:spawn [sync] Running "yarn create-react-app app-ui" +0ms

yarn run v1.22.4
$ C:\Users\Jacob\Documents\dev\test1\node_modules\.bin\create-react-app app-ui

Creating a new React app in C:\Users\Jacob\Documents\dev\test1\app-ui.

Installing packages. This might take a couple of minutes.
Installing react, react-dom, and react-scripts with cra-template...

[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@1.2.12: The platform "win32" is incompatible with this module.
info "fsevents@1.2.12" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.1.2: The platform "win32" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning "react-scripts > @typescript-eslint/eslint-plugin > tsutils@3.17.1" has unmet peer dependency "typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta".
[4/4] Building fresh packages...
success Saved lockfile.
success Saved 15 new dependencies.
info Direct dependencies
├─ cra-template@1.0.3
├─ react-dom@16.13.1
├─ react-scripts@3.4.1
└─ react@16.13.1
info All dependencies
├─ @babel/helper-member-expression-to-functions@7.11.0
├─ @babel/plugin-syntax-typescript@7.10.4
├─ @babel/plugin-transform-flow-strip-types@7.9.0
├─ @babel/plugin-transform-runtime@7.9.0
├─ @babel/plugin-transform-typescript@7.11.0
├─ @babel/preset-typescript@7.9.0
├─ babel-preset-react-app@9.1.2
├─ cra-template@1.0.3
├─ eslint-config-react-app@5.2.1
├─ react-dev-utils@10.2.1
├─ react-dom@16.13.1
├─ react-error-overlay@6.0.7
├─ react-scripts@3.4.1
├─ react@16.13.1
└─ scheduler@0.19.1

Initialized a git repository.

Installing template dependencies using yarnpkg...
[1/4] Resolving packages...
warning @testing-library/react > @types/testing-library__react > @types/testing-library__dom@7.5.0: This is a stub types definition. testing-library__dom provides its own type definitions, so you do not need this installed.
[2/4] Fetching packages...
info fsevents@2.1.2: The platform "win32" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.12: The platform "win32" is incompatible with this module.
info "fsevents@1.2.12" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning "react-scripts > @typescript-eslint/eslint-plugin > tsutils@3.17.1" has unmet peer dependency "typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta".
warning " > @testing-library/user-event@7.2.1" has unmet peer dependency "@testing-library/dom@>=5".
[4/4] Building fresh packages...
success Saved lockfile.
success Saved 22 new dependencies.
info Direct dependencies
├─ @testing-library/jest-dom@4.2.4
├─ @testing-library/react@9.5.0
├─ @testing-library/user-event@7.2.1
├─ react-dom@16.13.1
└─ react@16.13.1
info All dependencies
├─ @babel/runtime-corejs3@7.11.2
├─ @sheerun/mutationobserver-shim@0.3.3
├─ @testing-library/dom@6.16.0
├─ @testing-library/jest-dom@4.2.4
├─ @testing-library/react@9.5.0
├─ @testing-library/user-event@7.2.1
├─ @types/aria-query@4.2.0
├─ @types/prop-types@15.7.3
├─ @types/react-dom@16.9.8
├─ @types/react@16.9.45
├─ @types/testing-library__dom@7.5.0
├─ @types/testing-library__react@9.1.3
├─ aria-query@4.2.2
├─ css.escape@1.5.1
├─ csstype@3.0.2
├─ dom-accessibility-api@0.3.0
├─ min-indent@1.0.1
├─ react-dom@16.13.1
├─ react@16.13.1
├─ redent@3.0.0
├─ strip-indent@3.0.0
└─ wait-for-expect@3.0.2
Removing template package using yarnpkg...

[1/2] Removing module cra-template...
[2/2] Regenerating lockfile and installing missing dependencies...
info fsevents@2.1.2: The platform "win32" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.12: The platform "win32" is incompatible with this module.
info "fsevents@1.2.12" is an optional dependency and failed compatibility check. Excluding it from installation.
warning " > @testing-library/user-event@7.2.1" has unmet peer dependency "@testing-library/dom@>=5".
warning "react-scripts > @typescript-eslint/eslint-plugin > tsutils@3.17.1" has unmet peer dependency "typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta".
success Uninstalled packages.
Removing .git directory...

Success! Created app-ui at C:\Users\Jacob\Documents\dev\test1\app-ui
Inside that directory, you can run several commands:

  yarn start
    Starts the development server.

  yarn build
    Bundles the app into static files for production.

  yarn test
    Starts the test runner.

  yarn eject
    Removes this tool and copies build dependencies, configuration files
    and scripts into the app directory. If you do this, you can’t go back!

We suggest that you begin by typing:

  cd app-ui
  yarn start

Happy hacking!
Done in 99.87s.
 react-recipe
  Your installation completed.
  Run these commands to start:
    For development server (provides UI): cd app-ui; yarn start
    For tauri (runs desktop app) in this directory: yarn tauri dev
 +2m
```

zakpatterson committed:
Add `beforeXCommand` settings to recipes

This is so good, it makes it possible to just run:

> yarn init
> yarn add tauri
> yarn tauri init # with recipe selected, adds app ui
> yarn tauri dev # starts app ui dev server and rust component

zakpatterson reviewed (commented):
Renamed init to create, then created a new init command.

zakpatterson commented:
Ok, as of now starting a new project works as:

```
yarn init
yarn add tauri
yarn tauri create
yarn tauri dev
```

Anyone trying to add tauri to an existing project using `yarn tauri init` should notice no changes.

zakpatterson committed:
Update changes file

zakpatterson committed:
Rename init to create

Also improve so it exports a function that can be passed
arguments, instead of just reading directly from command line

zakpatterson committed:
Add new tauri init command aliased to create

zakpatterson committed:
Remove unneeded import

zakpatterson committed:
Add customized splash page for react recipes

zakpatterson commented:
![Peek 2020-08-11 11-39](https://user-images.githubusercontent.com/1987109/89924511-bc290080-dbc7-11ea-8d7b-015a1a625713.gif)

jbolda commented:
I ran the typescript recipe and getting the following error on `yarn tauri dev`.

```shell
❯ yarn tauri dev
yarn run v1.22.4
$ C:\Users\Jacob\Documents\dev\tauri\test-tauri-app\node_modules\.bin\tauri dev
[tauri]: running dev
 ERROR: tauri.conf.json > build has unknown property recipe +0ms
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Looks like we need to add this to the list of validation or scrub it from the conf.

zakpatterson commented:
Thanks @jbolda, I didn't consider that extra build setting and its impact on existing projects, I'll see if I can do without it

zakpatterson committed:
Remove unneeded recipe buildconfig property

zakpatterson commented:
Yep @jbolda it's much simpler without the added config property. Thanks for the catch.

jbolda commented on cli/tauri.js/src/api/recipes/react.ts:
```diff
@@ -0,0 +1,58 @@
+
+const afterCra = (): void => {
+  copyTemplates({
+    source: resolve(__dirname, '../../templates/recipes/react/'),
```
Does this work in both *nix and Windows?

zakpatterson replied:
Yep, same strategy as applied here: https://github.com/tauri-apps/tauri/blob/f208a68e40c804daf41d54539d3a5951679e8a64/cli/tauri.js/src/template/index.ts#L80-L92

jbolda replied:
Oh, I am getting the Tauri logo now, so you must have resolved that then!

jbolda commented on cli/tauri.js/src/api/recipes/react.ts:
```diff
@@ -0,0 +1,58 @@
+    ...cfg,
+    distDir: `../${uiAppDir}/build`,
+    devPath: 'http://localhost:3000',
+    beforeDevCommand: `yarn --cwd ${uiAppDir} start`,
```
I think we can specify a `BROWSER` env var here so it doesn't open the browser and the Tauri app.
https://create-react-app.dev/docs/advanced-configuration

zakpatterson replied:
22bcf3a

zakpatterson replied:
Nevermind, I reverted that as it's not going to work on windows. `beforeDevCommand` is executed using `exec` from `process`, which does spawn a shell and tries to run it.

So implementing this would require a change to the build config to allow `beforeDevCommand` to specify an environment variable, 

Something like 

```ts
interface Command {
  env: {[name: string]: string}
  command: string
}
interface TauriBuildConfig {
   beforeDevCommand: string | Command
}
```

This is not a big deal, but I think it lets the scope of this PR creep a little bit too much, I would prefer to do this in a follow-on PR if that's ok.

jbolda replied:
aye, agreed

zakpatterson replied:
Also, environment variables are passed into the child process, so for example `BROWSER=none yarn tauri dev` works on linux/osx, if that's the behavior the user wants.

zakpatterson committed:
Add no-browser setting to start react dev server

zakpatterson committed:
Revert "Add no-browser setting to start react dev server"

This reverts commit 22bcf3ac5a0474b206826417cf314a4be70324c5.

jbolda reviewed (approved):
Nothing else jumps out at me. I am sure we will have further tweaks, but I think we are ready to merge and let it morph organically from here. Thanks so much for your work on this! ❤️
